tooling/nargo_cli/src/cli/compile_cmd.rs

```diff
@@ -63,38 +64,53 @@ pub(crate) fn run(
     let circuit_dir = workspace.target_directory_path();
 
     let (np_language, is_opcode_supported) = backend.get_backend_info()?;
-    for package in &workspace {
-        // If `contract` package type, we're compiling every function in a 'contract' rather than just 'main'.
-        if package.is_contract() {
-            let (file_manager, compilation_result) = compile_contracts(
-                package,
-                &args.compile_options,
-                np_language,
-                &is_opcode_supported,
-            );
-            let contracts_with_debug_artifacts = report_errors(
-                compilation_result,
-                &file_manager,
-                args.compile_options.deny_warnings,
-            )?;
-
-            save_contracts(
-                contracts_with_debug_artifacts,
-                package,
-                &circuit_dir,
-                args.output_debug,
-            );
-        } else {
-            let (file_manager, compilation_result) =
-                compile_program(package, &args.compile_options, np_language, &is_opcode_supported);
-
-            let (program, debug_artifact) = report_errors(
-                compilation_result,
-                &file_manager,
-                args.compile_options.deny_warnings,
-            )?;
-            save_program(debug_artifact, program, package, &circuit_dir, args.output_debug);
-        }
+
+    let (binary_packages, contract_packages): (Vec<_>, Vec<_>) = workspace
+        .members
+        .iter()
+        .filter(|package| !package.is_library())
+        .partition(|package| package.is_binary());
+
+    // Compile all of the packages in parallel.
+    let program_results: Vec<(FileManager, CompilationResult<(CompiledProgram, DebugArtifact)>)> =
+        binary_packages
+            .par_iter()
+            .map(|package| {
+                compile_program(package, &args.compile_options, np_language, &is_opcode_supported)
+            })
+            .collect();
+    let contract_results: Vec<(
+        FileManager,
+        CompilationResult<Vec<(CompiledContract, DebugArtifact)>>,
+    )> = contract_packages
+        .par_iter()
+        .map(|package| {
+            compile_contracts(package, &args.compile_options, np_language, &is_opcode_supported)
+        })
+        .collect();
```

Comment:
Note that we still need to wait for all programs to compile before we compile contracts but this is a decent improvement on the current situation and we can iterate on this.